shared/util.go

```diff
@@ -728,32 +728,57 @@ func RunningInUserNS() bool {
 
 // ValidHostname checks the string is valid DNS hostname.
 func ValidHostname(name string) error {
+	match, err := validName(name, "[-a-zA-Z0-9]")
+	if err != nil {
+		return err
+	}
+
+	if !match {
+		return fmt.Errorf("Name can only contain alphanumeric and hyphen characters")
+	}
+
+	return nil
+}
+
+// ValidHostnameWithUnderscore performs the same checks as ValidHostname but also allows underscores.
+func ValidHostnameWithUnderscore(name string) error {
```

Comment:
That function name ends up being a bit of a lie then as per RFC, a hostname cannot be valid when containing an underscore ;)

Reply:
I've changed this in #9946 :)